Provide endpoint to list spaces in which a user has a role

xcoulon opened the pull request:
Added endpoint in `UserController`
Added tests in `user_blackbox_test.go` using go-vcr
to handle requests/responses with remote auth service.

fixes #2241
fixes fabric8-services/fabric8-auth#617
fixes openshiftio/openshift.io#3760

Signed-off-by: Xavier Coulon <xcoulon@redhat.com>

alexeykazakov commented on controller/user.go:
```diff
@@ -105,15 +112,93 @@ func convertToAppUser(request *goa.RequestData, user *authservice.User) *app.Use
+	if err != nil {
+		return jsonapi.JSONErrorResponse(ctx, err)
+	}
+	res, err := client.ListResourcesUser(goasupport.ForwardContextRequestID(ctx), authservice.ListResourcesUserPath(), "spaces")
```
Why "spaces"?
1. Why is it plural?
2. We should use the same resource type name which used by Auth in other API and internally which is "openshift.io/resource/space" currently. If we think that this resource type name is too long then it should be changed in the Auth service. Why we use different resource type names for the same resource types in different Auth API? This is confusing.

To be clear. It makes sense to introduce `/api/user/spaces` endpoint to WIT. What I don't like is the Auth endpoint: `/api/user/resources?type="spaces"` 

cc: @sbose78 @sbryzak

xcoulon replied:
@alexeykazakov so you would prefer something like `/api/user/resources?type=openshift.io/resource/space` in auth ? (I need to see if we can avoid surrounding `openshift.io/resource/space` with double quotes)

xcoulon replied:
(I've opened https://github.com/fabric8-services/fabric8-auth/issues/623 for this change)

sbose78 replied:
Agreed. @alexeykazakov is right - my bad. 

Yeah resource type name feels a bit long actually - but nevertheless, let's stick to the exact resource type name `/api/user/resources?type=openshift.io/resource/space` should work without quotes - just like `/api/login?redirect=https://auth/api/status` . 

You might consider encoding the param value it though.

alexeykazakov commented on controller/user.go:
```diff
@@ -33,20 +39,21 @@ type UserControllerConfiguration interface {
 // Show returns the authorized user based on the provided Token
 func (c *UserController) Show(ctx *app.ShowUserContext) error {
-	client, err := c.createClient(ctx)
+	client, err := newAuthClient(ctx, c.config, c.httpOptions...)
```
Well... it's not required for this PR but this Show method could be cleaned up.. There is no need to use Auth client here. Pure proxy will work just fine (see how it's done in /users controller). Originally we didn't switch to the proxy for /user because we need to call tenant init() here. But it's now done in Auth anyway. So, the whole Show() method could be just:

```
func (c *UsersController) Show(ctx *app.ShowUserContext) error {
	return proxy.RouteHTTP(ctx, c.config.GetAuthShortServiceHostName())
}
```

xcoulon replied:
good point, but I'd rather open another PR for that.

alexeykazakov commented on design/account.go:
```diff
@@ -94,6 +94,35 @@ var userDataAttributes = a.Type("UserDataAttributes", func() {
+// userSpacesDataAttributes contains info about the role and scopes that the user has in the resource
+var userSpacesDataAttributes = a.Type("UserSpacesDataAttributes", func() {
+	a.Attribute("name", d.String, "The name of the space")
+	a.Attribute("role", d.String, "The role of the user in the corresponding space")
```
Right now we don't use that but in the future the user can have multiple roles in the space. I wonder if we should reflect it's in API right now.

xcoulon replied:
good point, and `auth` should also better handle this case. I'll work on it as part of https://github.com/fabric8-services/fabric8-auth/issues/622

alexeykazakov commented on design/account.go:
```diff
@@ -94,6 +94,35 @@ var userDataAttributes = a.Type("UserDataAttributes", func() {
+
+// userSpacesData represents a resource for which a user has a role
+var userSpacesData = a.Type("UserSpacesData", func() {
+	a.Attribute("id", d.String, "id of the resource that in which the user has a role")
```
It's not just some resource. It's always the ID of the space. IMO it's worth being more specific here.

xcoulon replied:
right, let me fix that.

xcoulon commented:
thanks for the review @alexeykazakov! I now have a couple of changes to do in `auth` in order to handle your comments :D

alexeykazakov commented:
Thanks @xcoulon! I've changed the PR to WIP because it now depends on upcoming changes in Auth.

xcoulon commented:
@alexeykazakov I've updated the PR after the endpoint was refactored in `auth` (not returning roles and scopes any longer). I've also removed the `WIP` prefix on the PR title ;)
Could you please let me know if you have further concerns/comments on this PR, please?

xcoulon commented:
[test]

xcoulon commented:
[test]

xcoulon commented on controller/user.go:
```diff
@@ -105,15 +112,91 @@ func convertToAppUser(request *goa.RequestData, user *authservice.User) *app.Use
+	if err != nil {
+		return jsonapi.JSONErrorResponse(ctx, err)
+	}
+	res, err := client.ListResourcesUser(goasupport.ForwardContextRequestID(ctx), authservice.ListResourcesUserPath(), "spaces")
```
ouch! Yes!

xcoulon replied:
@alexeykazakov this has been fixed in 10fa333c. Thanks for spotting this error!

xcoulon commented:
@alexeykazakov this is fixed in 10fa333c

sbose78 commented on controller/user_blackbox_test.go:
```diff
@@ -0,0 +1,156 @@
+
+	s.T().Run("ok", func(t *testing.T) {
+
+		t.Run("user has a role in 1 space", func(t *testing.T) {
```
Let's have a test for 'user doesn't have a role in any space'?

xcoulon replied:
@sbose78 sure, let me add that

xcoulon replied:
@sbose78 I've just added the test in 616ae896

xcoulon commented:
thanks for the review, @alexeykazakov and @sbose78 ! I'm going to merge the latest changes from `master` then merge this PR afterwards.
